File: tools/R/analyze_results.R
#!/usr/bin/env Rscript

args <- commandArgs(trailingOnly = TRUE)
root_dir <- if (length(args) >= 1) args[[1]] else "."
pdr_threshold <- if (length(args) >= 2) as.numeric(args[[2]]) else 0.9

summary_path <- file.path(root_dir, "results", "summary.csv")
analysis_dir <- file.path(root_dir, "results", "analysis")

if (!file.exists(summary_path)) stop("summary.csv not found: ", summary_path)
if (!dir.exists(analysis_dir)) dir.create(analysis_dir, recursive = TRUE)

summary <- read.csv(summary_path, stringsAsFactors = FALSE)

numeric_cols <- c(
  "n_senders","seed","success_ratio","interference_ratio","send_interval_s",
  "rx_count","tx_expected","pdr","avg_delay_ms","p95_delay_ms",
  "dio_count","dao_count","duration_s","warmup_s","measure_s"
)
for (col in numeric_cols) {
  if (col %in% names(summary)) summary[[col]] <- suppressWarnings(as.numeric(summary[[col]]))
}

summary$overhead_per_s <- (summary$dio_count + summary$dao_count) / summary$duration_s

cond_cols <- c("mode","stage","n_senders","success_ratio","interference_ratio","send_interval_s")
agg <- aggregate(
  cbind(pdr, avg_delay_ms, p95_delay_ms, rx_count, tx_expected, dio_count, dao_count) ~ .,
  data = summary[, c(cond_cols, "pdr","avg_delay_ms","p95_delay_ms","rx_count","tx_expected","dio_count","dao_count")],
  FUN = mean,
  na.rm = TRUE
)
runs <- aggregate(
  seed ~ .,
  data = summary[, c(cond_cols, "seed")],
  FUN = length
)
names(runs)[names(runs) == "seed"] <- "runs"
agg <- merge(agg, runs, all.x = TRUE)

agg$collapse_flag <- (agg$pdr < pdr_threshold) | (agg$rx_count <= 0)

stage1 <- agg[agg$stage == "stage1", , drop = FALSE]
stage1_rows <- list()
for (mode in sort(unique(stage1$mode))) {
  subset_mode <- stage1[stage1$mode == mode, , drop = FALSE]
  subset_mode <- subset_mode[order(subset_mode$n_senders), ]
  idx <- which(subset_mode$collapse_flag)
  if (length(idx) > 0) {
    row <- subset_mode[idx[1], c("mode","stage","n_senders","pdr","avg_delay_ms","p95_delay_ms","runs")]
  } else {
    row <- data.frame(
      mode = mode, stage = "stage1", n_senders = NA,
      pdr = NA, avg_delay_ms = NA, p95_delay_ms = NA, runs = NA
    )
  }
  stage1_rows[[length(stage1_rows) + 1]] <- row
}
stage1_out <- do.call(rbind, stage1_rows)
write.csv(stage1_out, file.path(analysis_dir, "collapse_stage1.csv"), row.names = FALSE)

stage2 <- agg[agg$stage == "stage2", , drop = FALSE]
stage2_rows <- list()
if (nrow(stage2) > 0) {
  key <- interaction(stage2$mode, stage2$n_senders, stage2$interference_ratio, drop = TRUE)
  groups <- split(stage2, key)
  for (group in groups) {
    group <- group[order(-group$success_ratio), ]
    idx <- which(group$collapse_flag)
    if (length(idx) > 0) {
      row <- group[idx[1], c("mode","stage","n_senders","success_ratio","interference_ratio","pdr","avg_delay_ms","p95_delay_ms","runs")]
      stage2_rows[[length(stage2_rows) + 1]] <- row
    }
  }
}
stage2_out <- if (length(stage2_rows) > 0) do.call(rbind, stage2_rows) else data.frame()
write.csv(stage2_out, file.path(analysis_dir, "collapse_stage2.csv"), row.names = FALSE)

stage3 <- agg[agg$stage == "stage3", , drop = FALSE]
stage3_rows <- list()
if (nrow(stage3) > 0) {
  key <- interaction(stage3$mode, stage3$n_senders, stage3$success_ratio, stage3$interference_ratio, drop = TRUE)
  groups <- split(stage3, key)
  for (group in groups) {
    group <- group[order(group$send_interval_s), ]
    idx <- which(group$collapse_flag)
    if (length(idx) > 0) {
      row <- group[idx[1], c("mode","stage","n_senders","success_ratio","interference_ratio","send_interval_s","pdr","avg_delay_ms","p95_delay_ms","runs")]
      stage3_rows[[length(stage3_rows) + 1]] <- row
    }
  }
}
stage3_out <- if (length(stage3_rows) > 0) do.call(rbind, stage3_rows) else data.frame()
write.csv(stage3_out, file.path(analysis_dir, "collapse_stage3.csv"), row.names = FALSE)

comp_mean <- aggregate(
  cbind(pdr, avg_delay_ms, p95_delay_ms, overhead_per_s) ~ mode + stage,
  data = summary,
  FUN = mean,
  na.rm = TRUE
)
comp_median <- aggregate(
  cbind(pdr, avg_delay_ms, p95_delay_ms, overhead_per_s) ~ mode + stage,
  data = summary,
  FUN = median,
  na.rm = TRUE
)
names(comp_mean)[3:ncol(comp_mean)] <- paste0(names(comp_mean)[3:ncol(comp_mean)], "_mean")
names(comp_median)[3:ncol(comp_median)] <- paste0(names(comp_median)[3:ncol(comp_median)], "_median")
comp <- merge(comp_mean, comp_median, by = c("mode","stage"))
runs_by_mode <- aggregate(seed ~ mode + stage, data = summary, FUN = length)
names(runs_by_mode)[names(runs_by_mode) == "seed"] <- "runs"
comp <- merge(comp, runs_by_mode, by = c("mode","stage"))
write.csv(comp, file.path(analysis_dir, "mode_stage_comparison.csv"), row.names = FALSE)

docs_dir <- file.path(root_dir, "docs")
tables_dir <- file.path(docs_dir, "tables")
figures_dir <- file.path(docs_dir, "figures")
if (!dir.exists(tables_dir)) dir.create(tables_dir, recursive = TRUE)
if (!dir.exists(figures_dir)) dir.create(figures_dir, recursive = TRUE)

stage1_summary <- summary[summary$stage == "stage1", , drop = FALSE]
if (nrow(stage1_summary) > 0) {
  stage1_summary$pdr_pct <- stage1_summary$pdr * 100
  agg_stage1 <- aggregate(
    cbind(pdr_pct, avg_delay_ms) ~ mode + n_senders,
    data = stage1_summary,
    FUN = function(x) c(mean = mean(x, na.rm = TRUE), sd = sd(x, na.rm = TRUE))
  )
  agg_stage1 <- do.call(data.frame, agg_stage1)
  names(agg_stage1) <- c(
    "mode", "n_senders",
    "mean_pdr", "sd_pdr",
    "mean_delay", "sd_delay"
  )
  runs_stage1 <- aggregate(seed ~ mode + n_senders, data = stage1_summary, FUN = length)
  names(runs_stage1)[names(runs_stage1) == "seed"] <- "n_runs"
  agg_stage1 <- merge(agg_stage1, runs_stage1, by = c("mode", "n_senders"))

  tex_path <- file.path(tables_dir, "stage1_summary.tex")
  lines <- c(
    "\\begin{table}[H]",
    "\\centering",
    "\\caption{Performance Comparison of RPL Variants (Stage 1)}",
    "\\label{tab:stage1_results}",
    "\\begin{tabular}{llrrr}",
    "\\toprule",
    "Mode & N Senders & PDR (\\%) & Delay (ms) & Runs \\\\",
    "\\midrule"
  )
  mode_order <- c("rpl-classic", "rpl-lite", "brpl")
  for (mode in mode_order) {
    mode_rows <- agg_stage1[agg_stage1$mode == mode, , drop = FALSE]
    if (nrow(mode_rows) == 0) next
    mode_rows <- mode_rows[order(mode_rows$n_senders), ]
    for (i in seq_len(nrow(mode_rows))) {
      row <- mode_rows[i, ]
      pdr_str <- sprintf("%.2f $\\\\pm$ %.2f", row$mean_pdr, row$sd_pdr)
      delay_str <- sprintf("%.2f $\\\\pm$ %.2f", row$mean_delay, row$sd_delay)
      lines <- c(lines, sprintf(
        "%s & %d & %s & %s & %d \\\\",
        row$mode, row$n_senders, pdr_str, delay_str, row$n_runs
      ))
    }
  }
  lines <- c(lines, "\\bottomrule", "\\end{tabular}", "\\end{table}")
  writeLines(lines, tex_path)

  plot_with_error <- function(df, y_mean, y_sd, ylab, title, out_path) {
    pdf(out_path)
    mode_order <- c("rpl-classic", "rpl-lite", "brpl")
    colors <- c("rpl-classic" = "#1b9e77", "rpl-lite" = "#d95f02", "brpl" = "#7570b3")
    plot(NULL,
      xlim = range(df$n_senders, na.rm = TRUE),
      ylim = c(0, max(df[[y_mean]] + df[[y_sd]], na.rm = TRUE) * 1.1),
      xlab = "Number of Senders",
      ylab = ylab,
      main = title
    )
    for (mode in mode_order) {
      mode_rows <- df[df$mode == mode, , drop = FALSE]
      if (nrow(mode_rows) == 0) next
      mode_rows <- mode_rows[order(mode_rows$n_senders), ]
      x <- mode_rows$n_senders
      y <- mode_rows[[y_mean]]
      ysd <- mode_rows[[y_sd]]
      lines(x, y, type = "b", pch = 19, col = colors[mode])
      segments(x, y - ysd, x, y + ysd, col = colors[mode])
    }
    legend("topleft", legend = mode_order, col = colors[mode_order], lty = 1, pch = 19, bty = "n")
    dev.off()
  }

  plot_with_error(
    agg_stage1, "mean_pdr", "sd_pdr",
    "PDR (%)", "Packet Delivery Ratio by Number of Senders",
    file.path(figures_dir, "stage1_pdr.pdf")
  )
  plot_with_error(
    agg_stage1, "mean_delay", "sd_delay",
    "Delay (ms)", "Average Delay by Number of Senders",
    file.path(figures_dir, "stage1_delay.pdf")
  )
}

cat("Analysis complete.\n")
cat("Outputs:\n")
cat(" -", file.path(analysis_dir, "collapse_stage1.csv"), "\n")
cat(" -", file.path(analysis_dir, "collapse_stage2.csv"), "\n")
cat(" -", file.path(analysis_dir, "collapse_stage3.csv"), "\n")
cat(" -", file.path(analysis_dir, "mode_stage_comparison.csv"), "\n")

File: tools/python/analyze_results.py
#!/usr/bin/env python3
"""
analyze_results.py
RPL 실험 결과를 분석하고 LaTeX 테이블을 생성합니다.
"""

import os
import glob
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from pathlib import Path

# 한글 폰트 설정
plt.rcParams['font.family'] = 'NanumGothic'
plt.rcParams['axes.unicode_minus'] = False

def parse_filename(filename):
    """파일명에서 메타데이터 추출"""
    parts = Path(filename).stem.split('_')
    return {
        'n_senders': int(parts[0].replace('N', '')),
        'seed': int(parts[1].replace('seed', '')),
        'success_ratio': float(parts[2].replace('sr', '').replace('p', '.')),
        'interference_ratio': float(parts[3].replace('ir', '').replace('p', '.')),
        'send_interval': int(parts[4].replace('si', ''))
    }

def read_result_csv(filepath, mode):
    """CSV 파일 읽기"""
    try:
        df = pd.read_csv(filepath)
        metadata = parse_filename(filepath)
        for key, value in metadata.items():
            df[key] = value
        df['mode'] = mode
        return df
    except Exception as e:
        print(f"Error reading {filepath}: {e}")
        return None

def read_all_results(stage='stage1'):
    """모든 결과 파일 읽기"""
    results_dir = Path('/home/dev/WSN-IoT-lab/rpl-benchmark/results/raw') / stage
    all_data = []
    
    for mode in ['rpl-classic', 'rpl-lite', 'brpl']:
        mode_dir = results_dir / mode
        if not mode_dir.exists():
            print(f"Directory not found: {mode_dir}")
            continue
            
        csv_files = glob.glob(str(mode_dir / '*.csv'))
        print(f"Found {len(csv_files)} files for {mode}")
        
        for csv_file in csv_files:
            df = read_result_csv(csv_file, mode)
            if df is not None:
                all_data.append(df)
    
    if not all_data:
        print("No data found!")
        return None
    
    return pd.concat(all_data, ignore_index=True)

def generate_summary(data):
    """통계 요약 생성"""
    summary = data.groupby(['mode', 'n_senders', 'seed']).agg({
        'seq': 'max',
        'delay_ms': ['mean', 'median', 'max', 'count']
    }).reset_index()
    
    summary.columns = ['mode', 'n_senders', 'seed', 'total_sent', 
                       'avg_delay_ms', 'median_delay_ms', 'max_delay_ms', 'total_received']
    
    summary['pdr'] = (summary['total_received'] / summary['total_sent']) * 100
    
    return summary

def aggregate_by_mode(summary):
    """모드별 평균 통계"""
    aggregated = summary.groupby(['mode', 'n_senders']).agg({
        'pdr': ['mean', 'std', 'count'],
        'avg_delay_ms': ['mean', 'std']
    }).reset_index()
    
    aggregated.columns = ['mode', 'n_senders', 'mean_pdr', 'sd_pdr', 'n_runs', 
                          'mean_delay', 'sd_delay']
    
    return aggregated

def generate_latex_table(aggregated, output_file):
    """LaTeX 테이블 생성"""
    lines = []
    lines.append("\\begin{table}[H]")
    lines.append("\\centering")
    lines.append("\\caption{Performance Comparison of RPL Variants (Stage 1)}")
    lines.append("\\label{tab:stage1_results}")
    lines.append("\\begin{tabular}{llrrr}")
    lines.append("\\toprule")
    lines.append("Mode & N Senders & PDR (\\%) & Delay (ms) & Runs \\\\")
    lines.append("\\midrule")
    
    for mode in ['rpl-classic', 'rpl-lite', 'brpl']:
        mode_data = aggregated[aggregated['mode'] == mode].sort_values('n_senders')
        for _, row in mode_data.iterrows():
            pdr_str = f"{row['mean_pdr']:.2f} $\\pm$ {row['sd_pdr']:.2f}"
            delay_str = f"{row['mean_delay']:.2f} $\\pm$ {row['sd_delay']:.2f}"
            lines.append(f"{mode} & {row['n_senders']} & {pdr_str} & {delay_str} & {int(row['n_runs'])} \\\\")
    
    lines.append("\\bottomrule")
    lines.append("\\end{tabular}")
    lines.append("\\end{table}")
    
    with open(output_file, 'w') as f:
        f.write('\n'.join(lines))
    
    print(f"LaTeX table saved to {output_file}")

def plot_pdr_comparison(aggregated, output_file):
    """PDR 비교 그래프"""
    plt.figure(figsize=(10, 6))
    
    for mode in ['rpl-classic', 'rpl-lite', 'brpl']:
        mode_data = aggregated[aggregated['mode'] == mode].sort_values('n_senders')
        plt.errorbar(mode_data['n_senders'], mode_data['mean_pdr'], 
                     yerr=mode_data['sd_pdr'], marker='o', label=mode, 
                     linewidth=2, markersize=8, capsize=5)
    
    plt.xlabel('Number of Senders', fontsize=12, fontweight='bold')
    plt.ylabel('PDR (%)', fontsize=12, fontweight='bold')
    plt.title('Packet Delivery Ratio by Number of Senders', fontsize=14, fontweight='bold')
    plt.legend(fontsize=11)
    plt.grid(True, alpha=0.3)
    plt.ylim(0, 105)
    plt.tight_layout()
    plt.savefig(output_file, dpi=300, bbox_inches='tight')
    print(f"Plot saved to {output_file}")
    plt.close()

def plot_delay_comparison(aggregated, output_file):
    """Delay 비교 그래프"""
    plt.figure(figsize=(10, 6))
    
    for mode in ['rpl-classic', 'rpl-lite', 'brpl']:
        mode_data = aggregated[aggregated['mode'] == mode].sort_values('n_senders')
        plt.errorbar(mode_data['n_senders'], mode_data['mean_delay'], 
                     yerr=mode_data['sd_delay'], marker='o', label=mode, 
                     linewidth=2, markersize=8, capsize=5)
    
    plt.xlabel('Number of Senders', fontsize=12, fontweight='bold')
    plt.ylabel('Delay (ms)', fontsize=12, fontweight='bold')
    plt.title('Average Delay by Number of Senders', fontsize=14, fontweight='bold')
    plt.legend(fontsize=11)
    plt.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.savefig(output_file, dpi=300, bbox_inches='tight')
    print(f"Plot saved to {output_file}")
    plt.close()

def main():
    print("=" * 60)
    print("RPL Benchmark Results Analysis")
    print("=" * 60)
    print()
    
    # 데이터 읽기
    print("Reading experiment results...")
    data = read_all_results('stage1')
    
    if data is None or len(data) == 0:
        print("No data to analyze!")
        return
    
    print(f"Total rows: {len(data)}")
    print(f"Modes found: {data['mode'].unique()}")
    print(f"Node counts: {sorted(data['n_senders'].unique())}")
    print()
    
    # 통계 요약
    print("Generating summary statistics...")
    summary = generate_summary(data)
    aggregated = aggregate_by_mode(summary)
    
    print("\nAggregated Results:")
    print(aggregated.to_string(index=False))
    print()
    
    # 출력 디렉토리 생성
    docs_dir = Path('/home/dev/WSN-IoT-lab/rpl-benchmark/docs')
    tables_dir = docs_dir / 'tables'
    figures_dir = docs_dir / 'figures'
    tables_dir.mkdir(parents=True, exist_ok=True)
    figures_dir.mkdir(parents=True, exist_ok=True)
    
    # LaTeX 테이블 생성
    print("Generating LaTeX table...")
    generate_latex_table(aggregated, tables_dir / 'stage1_summary.tex')
    
    # 그래프 생성
    print("Generating plots...")
    plot_pdr_comparison(aggregated, figures_dir / 'stage1_pdr.pdf')
    plot_delay_comparison(aggregated, figures_dir / 'stage1_delay.pdf')
    
    print()
    print("=" * 60)
    print("Analysis complete!")
    print("=" * 60)

if __name__ == '__main__':
    main()
